File: PRACTICA 2/complejidad.py
import time
import tracemalloc
import random
import sys

def ejecutar_todas_las_pruebas_python():
    """
    Funcion principal para ejecutar todas las pruebas de algoritmos en Python.
    """
    # Aumentar el limite de recursion para Merge Sort con n=10^5
    # Es necesario para evitar un RecursionError con arreglos grandes.
    sys.setrecursionlimit(2000)

    # --- 1. Implementacion de Busqueda Lineal ---
    def busqueda_lineal(arreglo, objetivo):
        for indice in range(len(arreglo)):
            if arreglo[indice] == objetivo:
                return indice
        return -1

    # --- 2. Implementacion de Busqueda Binaria ---
    def busqueda_binaria(arreglo, objetivo):
        bajo, alto = 0, len(arreglo) - 1
        while bajo <= alto:
            medio = (bajo + alto) // 2
            if arreglo[medio] == objetivo:
                return medio
            elif arreglo[medio] < objetivo:
                bajo = medio + 1
            else:
                alto = medio - 1
        return -1

    # --- 3. Implementacion de Ordenamiento Burbuja ---
    def ordenamiento_burbuja(arreglo):
        n = len(arreglo)
        for i in range(n):
            intercambio = False
            for j in range(0, n-i-1):
                if arreglo[j] > arreglo[j+1]:
                    arreglo[j], arreglo[j+1] = arreglo[j+1], arreglo[j]
                    intercambio = True
            if not intercambio:
                break
        return arreglo

    # --- 4. Implementacion de Ordenamiento por Mezcla ---
    def ordenamiento_mezcla(arreglo):
        if len(arreglo) > 1:
            medio = len(arreglo) // 2
            izquierda = arreglo[:medio]
            derecha = arreglo[medio:]
            
            ordenamiento_mezcla(izquierda)
            ordenamiento_mezcla(derecha)
            
            i = j = k = 0
            while i < len(izquierda) and j < len(derecha):
                if izquierda[i] < derecha[j]:
                    arreglo[k] = izquierda[i]; i += 1
                else:
                    arreglo[k] = derecha[j]; j += 1
                k += 1
            
            while i < len(izquierda):
                arreglo[k] = izquierda[i]; i += 1; k += 1
            
            while j < len(derecha):
                arreglo[k] = derecha[j]; j += 1; k += 1
        return arreglo

    # --- 5. Implementacion de Fibonacci Recursivo ---
    def fibonacci_recursivo(n):
        if n <= 1:
            return n
        return fibonacci_recursivo(n-1) + fibonacci_recursivo(n-2)


    # Pruebas de Busqueda Lineal
    print("\n--- 1. Busqueda Lineal ---")
    print(f"{'Tamano (n)':<15} | {'Tiempo (ms)':<15} | {'Pico Memoria (KiB)':<20}")
    print("-" * 55)
    for n in [10**3, 10**4, 10**5]:
        arreglo = list(range(n)); objetivo = n - 1 
        inicio_tiempo = time.perf_counter()
        busqueda_lineal(arreglo, objetivo)
        fin_tiempo = time.perf_counter()
        tiempo_transcurrido = (fin_tiempo - inicio_tiempo) * 1000
        tracemalloc.start()
        busqueda_lineal(arreglo, objetivo)
        _, pico = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        print(f"{n:<15,} | {tiempo_transcurrido:<15.4f} | {pico / 1024:<20.4f}")
    
    # Pruebas de Busqueda Binaria
    print("\n--- 2. Busqueda Binaria ---")
    print(f"{'Tamano (n)':<15} | {'Tiempo (ms)':<15} | {'Pico Memoria (KiB)':<20}")
    print("-" * 55)
    for n in [10**3, 10**4, 10**5]:
        arreglo = list(range(n)); objetivo = n - 1
        inicio_tiempo = time.perf_counter()
        busqueda_binaria(arreglo, objetivo)
        fin_tiempo = time.perf_counter()
        tiempo_transcurrido = (fin_tiempo - inicio_tiempo) * 1000
        tracemalloc.start()
        busqueda_binaria(arreglo, objetivo)
        _, pico = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        print(f"{n:<15,} | {tiempo_transcurrido:<15.4f} | {pico / 1024:<20.4f}")

    # Pruebas de Ordenamiento Burbuja
    print("\n--- 3. Ordenamiento por Burbuja ---")
    print(f"{'Tamano (n)':<15} | {'Tiempo (ms)':<15} | {'Pico Memoria (KiB)':<20}")
    print("-" * 55)
    for n in [10**3, 10**4]: # 10^5 es demasiado lento
        arreglo = list(range(n, 0, -1))
        inicio_tiempo = time.perf_counter()
        ordenamiento_burbuja(arreglo.copy())
        fin_tiempo = time.perf_counter()
        tiempo_transcurrido = (fin_tiempo - inicio_tiempo) * 1000
        tracemalloc.start()
        ordenamiento_burbuja(arreglo.copy())
        _, pico = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        print(f"{n:<15,} | {tiempo_transcurrido:<15.4f} | {pico / 1024:<20.4f}")

    # Pruebas de Ordenamiento por Mezcla
    print("\n--- 4. Ordenamiento por Mezcla ---")
    print(f"{'Tamano (n)':<15} | {'Tiempo (ms)':<15} | {'Pico Memoria (KiB)':<20}")
    print("-" * 55)
    for n in [10**3, 10**4, 10**5]:
        arreglo = [random.randint(0, n) for _ in range(n)]
        inicio_tiempo = time.perf_counter()
        ordenamiento_mezcla(arreglo.copy())
        fin_tiempo = time.perf_counter()
        tiempo_transcurrido = (fin_tiempo - inicio_tiempo) * 1000
        tracemalloc.start()
        ordenamiento_mezcla(arreglo.copy())
        _, pico = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        print(f"{n:<15,} | {tiempo_transcurrido:<15.4f} | {pico / 1024:<20.4f}")

    # Pruebas de Fibonacci Recursivo
    print("\n--- 5. Fibonacci Recursivo ---")
    print(f"{'Valor (n)':<15} | {'Tiempo (ms)':<15} | {'Pico Memoria (KiB)':<20}")
    print("-" * 55)
    for n in range(1, 21):
        inicio_tiempo = time.perf_counter()
        fibonacci_recursivo(n)
        fin_tiempo = time.perf_counter()
        tiempo_transcurrido = (fin_tiempo - inicio_tiempo) * 1000
        tracemalloc.start()
        fibonacci_recursivo(n)
        _, pico = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        # Imprimir solo una muestra para no saturar la salida
        if n <= 5 or n % 5 == 0:
            print(f"{n:<15} | {tiempo_transcurrido:<15.4f} | {pico / 1024:<20.4f}")

# --- Punto de Entrada del Script ---
if __name__ == "__main__":
    ejecutar_todas_las_pruebas_python()

File: PRACTICA 2/complejidad.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <locale.h>

// --- 1. Implementacion de Busqueda Lineal ---
int busqueda_lineal(int arreglo[], int tam, int objetivo) {
    for (int i = 0; i < tam; i++) {
        if (arreglo[i] == objetivo) return i;
    }
    return -1;
}

// --- 2. Implementacion de Busqueda Binaria ---
int busqueda_binaria(int arreglo[], int tam, int objetivo) {
    int izq = 0, der = tam - 1;
    while (izq <= der) {
        int mid = izq + (der - izq) / 2;
        if (arreglo[mid] == objetivo) return mid;
        if (arreglo[mid] < objetivo) izq = mid + 1;
        else der = mid - 1;
    }
    return -1;
}

// --- 3. Implementacion de Ordenamiento Burbuja ---
void intercambiar(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void ordenamiento_burbuja(int arreglo[], int tam) {
    for (int i = 0; i < tam - 1; i++) {
        int intercambio = 0;
        for (int j = 0; j < tam - i - 1; j++) {
            if (arreglo[j] > arreglo[j + 1]) {
                intercambiar(&arreglo[j], &arreglo[j + 1]);
                intercambio = 1;
            }
        }
        if (intercambio == 0) break;
    }
}

// --- 4. Implementacion de Ordenamiento por merge ---
void merge(int arreglo[], int izq, int mid, int der) {
    int n1 = mid - izq + 1;
    int n2 = der - mid;
    int* I = (int*)malloc(n1 * sizeof(int));
    int* D = (int*)malloc(n2 * sizeof(int));

    for (int i = 0; i < n1; i++) I[i] = arreglo[izq + i];
    for (int j = 0; j < n2; j++) D[j] = arreglo[mid + 1 + j];

    int i = 0, j = 0, k = izq;
    while (i < n1 && j < n2) {
        if (I[i] <= D[j]) arreglo[k++] = I[i++];
        else arreglo[k++] = D[j++];
    }
    while (i < n1) arreglo[k++] = I[i++];
    while (j < n2) arreglo[k++] = D[j++];
    free(I);
    free(D);
}

void ordenamiento_merge(int arreglo[], int izq, int der) {
    if (izq < der) {
        int mid = izq + (der - izq) / 2;
        ordenamiento_merge(arreglo, izq, mid);
        ordenamiento_merge(arreglo, mid + 1, der);
        merge(arreglo, izq, mid, der);
    }
}

// --- 5. Implementacion de Fibonacci Recursivo ---
long long fibonacci_recursivo(int n) {
    if (n <= 1) return n;
    return fibonacci_recursivo(n - 1) + fibonacci_recursivo(n - 2);
}


// --- Bloque Principal de Pruebas ---
int main() {
    setlocale(LC_ALL, "");
    srand(time(NULL)); // Semilla para números aleatorios

    // Pruebas Busqueda Lineal
    printf("\n--- 1. Busqueda Lineal ---\n");
    printf("| %-12s | %-15s | %-20s |\n", "Tamano (n)", "Tiempo (ms)", "Memoria Heap (KiB)");
    printf("|--------------|-----------------|----------------------|\n");
    int tamanos_busqueda[] = {1000, 10000, 100000};
    for (int i = 0; i < 3; i++) {
        int tam = tamanos_busqueda[i];
        int* arreglo = (int*)malloc(tam * sizeof(int));
        for (int j = 0; j < tam; j++) arreglo[j] = j;
        clock_t inicio = clock();
        busqueda_lineal(arreglo, tam, tam - 1);
        clock_t fin = clock();
        double tiempo_ms = ((double)(fin - inicio) / CLOCKS_PER_SEC) * 1000.0;
        printf("| %-12d | %-15.4f | %-20.4f |\n", tam, tiempo_ms, (double)(tam*sizeof(int))/1024.0);
        free(arreglo);
    }

    // Pruebas Busqueda Binaria
    printf("\n--- 2. Busqueda Binaria ---\n");
    printf("| %-12s | %-15s | %-20s |\n", "Tamano (n)", "Tiempo (ms)", "Memoria Heap (KiB)");
    printf("|--------------|-----------------|----------------------|\n");
    for (int i = 0; i < 3; i++) {
        int tam = tamanos_busqueda[i];
        int* arreglo = (int*)malloc(tam * sizeof(int));
        for (int j = 0; j < tam; j++) arreglo[j] = j;
        clock_t inicio = clock();
        busqueda_binaria(arreglo, tam, tam-1);
        clock_t fin = clock();
        double tiempo_ms = ((double)(fin - inicio) / CLOCKS_PER_SEC) * 1000.0;
        printf("| %-12d | %-15.4f | %-20.4f |\n", tam, tiempo_ms, (double)(tam*sizeof(int))/1024.0);
        free(arreglo);
    }
    
    // Pruebas Ordenamiento Burbuja
    printf("\n--- 3. Ordenamiento por Burbuja ---\n");
    printf("| %-12s | %-15s | %-20s |\n", "Tamano (n)", "Tiempo (ms)", "Memoria Heap (KiB)");
    printf("|--------------|-----------------|----------------------|\n");
    int tamanos_burbuja[] = {1000, 10000, 100000};
    for (int i = 0; i < 3; i++) {
        int tam = tamanos_burbuja[i];
        int* arreglo = (int*)malloc(tam * sizeof(int));
        for (int j = 0; j < tam; j++) arreglo[j] = tam - j; // Peor caso
        clock_t inicio = clock();
        ordenamiento_burbuja(arreglo, tam);
        clock_t fin = clock();
        double tiempo_ms = ((double)(fin - inicio) / CLOCKS_PER_SEC) * 1000.0;
        printf("| %-12d | %-15.4f | %-20.4f |\n", tam, tiempo_ms, (double)(tam*sizeof(int))/1024.0);
        free(arreglo);
    }

    // Pruebas Ordenamiento por merge
    printf("\n--- 4. Ordenamiento por merge ---\n");
    printf("| %-12s | %-15s | %-20s |\n", "Tamano (n)", "Tiempo (ms)", "Memoria Heap (KiB)");
    printf("|--------------|-----------------|----------------------|\n");
    for (int i = 0; i < 3; i++) {
        int tam = tamanos_busqueda[i];
        int* arreglo = (int*)malloc(tam * sizeof(int));
        for (int j = 0; j < tam; j++) arreglo[j] = rand() % tam;
        clock_t inicio = clock();
        ordenamiento_merge(arreglo, 0, tam - 1);
        clock_t fin = clock();
        double tiempo_ms = ((double)(fin - inicio) / CLOCKS_PER_SEC) * 1000.0;
        printf("| %-12d | %-15.4f | %-20.4f |\n", tam, tiempo_ms, (double)(tam*sizeof(int))/1024.0);
        free(arreglo);
    }

    // Pruebas Fibonacci
    printf("\n--- 5. Fibonacci Recursivo ---\n");
    printf("| %-10s | %-15s | %-25s |\n", "Valor (n)", "Tiempo (ms)", "Memoria Stack (Teorico)");
    printf("|------------|-----------------|---------------------------|\n");
    for (int n = 1; n <= 20; n++) {
        clock_t inicio = clock();
        fibonacci_recursivo(n);
        clock_t fin = clock();
        double tiempo_ms = ((double)(fin - inicio) / CLOCKS_PER_SEC) * 1000.0;
        if (n <= 5 || n % 5 == 0) {
             printf("| %-10d | %-15.4f | Crece linealmente con n |\n", n, tiempo_ms);
        }
    }

    return 0;
}